### PR TITLE
Feature/36963 Helpdesk impersonation & home page

### PR DESCRIPTION
**Pull request by GeorgeChatzigiannis**
Introduced helpdesk site that includes impersonation form and helpdesk home page.
Helpdesk user can remove school session data by clicking on remove impersonation on home page.

Screenshot 2019-09-30 at 10 05 57

Screenshot 2019-09-30 at 10 07 47

### CODE REVIEW

File: admin/authentication/local-strategy.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ module.exports = async function (req, email, password, done) {
     // school with dfeNumber 1234567.
     // NOTE that this is not a generic implementation: you actually need to be called 'helpdesk' or 'service-manager'
     // rather than simply have the role.  This is fine - it's only for local-dev
-    const matches = email.match(/^(helpdesk|service-manager):(\d{7})$/)
+    const matches = email.match(/^(service-manager):(\d{7})$/)
```
Comment (GeorgeChatzigiannis) [inline admin/authentication/local-strategy.js]: @GuyHarwood @jon-shipley I removed the feature to directly impersonate a school from login page as this really serves no purpose now that the helpdesk site exists. Happy to discuss this further if needed.

File: admin/controllers/helpdesk.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+/**
+ * Renders the helpdesk (role) landing page
+ * @param {object} req
+ * @param {object} res
+ * @param {function} next
+ * @returns {Promise.<void>}
+ */
+controller.getHelpdeskHome = async (req, res, next) => {
```
Comment (GuyHarwood) [inline admin/controllers/helpdesk.js]: is this meant to be a helpdesk version of the school home page?  if so, i would expect us to simply use the existing one.

File: admin/authentication/local-strategy.js (hunk 1 above)
Comment (GuyHarwood) [reply admin/authentication/local-strategy.js]: makes sense

File: admin/controllers/helpdesk.js (hunk 2 above)
Comment (GeorgeChatzigiannis) [reply admin/controllers/helpdesk.js]: @GuyHarwood there are going to be additional features appearing on this home page.

Comment (GuyHarwood) [reply admin/controllers/helpdesk.js]: ok, so why don't we just have an optional partial for the PIN history link that is only rendered when its helpdesk?  that feels easier than duplicating all of the ongoing changes that we will make to the school homepage...

File: admin/services/school-impersonation.service.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+'use strict'
+const R = require('ramda')
+
+const schoolDataService = require('../services/data-access/school.data.service')
+const schoolImpersonationEmptyValueValidator = require('../lib/validator/school-impersonation/school-impersonation-empty-value-validator')
```
Comment (jon-shipley) [inline admin/services/school-impersonation.service.js]: I am not sure what an 'EmptyValue' validator is?  Isn't this the SchooImpersonatorFormValidator?  Why should the implementation be a part of the name?

File: admin/services/school-impersonation.service.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  let validationError
+  validationError = schoolImpersonationEmptyValueValidator.validate(dfeNumber)
+  if (validationError.hasError()) {
+    return validationError
+  }
+  let school
+  try {
+    school = await schoolDataService.sqlFindOneByDfeNumber(dfeNumber)
+  } catch (error) {}
+  validationError = schoolImpersonationDfeNumberValidator.validate(school)
```
Comment (jon-shipley) [inline admin/services/school-impersonation.service.js]: Why do we need two validators?

File: admin/services/school-impersonation.service.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+/**
+ * Remove impersonation data from request.user object
+ * @param {object} user
+ * @returns {void}
+ */
+schoolImpersonationService.removeImpersonation = (user) => {
+  const impersonationKeys = ['School', 'schoolId', 'timezone']
+  if (user.School || user.schoolId || user.timezone) {
```
Comment (jon-shipley) [inline admin/services/school-impersonation.service.js]: What is the point of the `if` condition here?  I suspect it is not needed as `delete` will always return true.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete

(unless deleting a property from the prototype chain; in which case it will throw)

File: admin/spec/back-end/lib/validator/school-impersonation/school-impersonation-empty-value-validator.spec.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
+'use strict'
+
+/* global describe, it expect */
+const schoolImpersonationEmptyValueValidator = require('../../../../../lib/validator/school-impersonation/school-impersonation-empty-value-validator')
+
+describe('schoolImpersonationEmptyValueValidator', function () {
+  it('should not add a validation error if dfeNumber is not empty', function () {
+    const dfeNumber = '1230000'
```
Comment (jon-shipley) [inline admin/spec/back-end/lib/validator/school-impersonation/school-impersonation-empty-value-validator.spec.js]: When do you convert the form input (string) into the required type for the DB lookup (number)?  Don't you also need to validate the post-conversion result to ensure it is an integer?

File: admin/controllers/helpdesk.js (hunk 2 above)
Comment (GeorgeChatzigiannis) [reply admin/controllers/helpdesk.js]: @GuyHarwood resolving this as it is now using the school home page

File: admin/authentication/local-strategy.js [hunk 7]
```diff
@@ -49,7 +49,13 @@ module.exports = async function (req, email, password, done) {
 
     const [school, role] = await Promise.all([schoolPromise, rolePromise])
 
-    if (!user || !role || !school) {
+    if (user && user.identifier !== 'helpdesk' && !school) {
```
[the remark this answers is not in the document]
Comment (GeorgeChatzigiannis) [reply admin/authentication/local-strategy.js]: @jon-shipley this has been significantly changed. Resolving this as I am now awaiting new feedback.

File: admin/controllers/helpdesk.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+ * Renders the helpdesk school impersonation page
+ * @param {object} req
+ * @param {object} res
+ * @param {function} next
+ * @param {object} error
+ * @returns {Promise.<void>}
+ */
+controller.getSchoolImpersonation = async (req, res, next, error = null) => {
+  res.locals.pageTitle = 'MTC Helpdesk School Impersonation'
+  schoolImpersonationService.removeImpersonation(req.user)
```
[the remark this answers is not in the document]
Comment (GeorgeChatzigiannis) [reply admin/controllers/helpdesk.js]: @jon-shipley this is to ensure the school session data are wiped. It is not a mandatory step as the form itself will populate the session with new school data however I liked the idea of having a clean session object instead of replacing existing key values.

File: admin/lib/errors/school-impersonation.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
+'use strict'
+
+module.exports = {
+  'noInput': 'Please type a dfe number',
+  'incorrectInput': 'The dfe number provided does not match a school'
```
[the remark this answers is not in the document]
Comment (GeorgeChatzigiannis) [reply admin/lib/errors/school-impersonation.js]: @jon-shipley indeed and it is updated.

File: admin/controllers/helpdesk.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+ * Submit helpdesk school impersonation
+ * @param {object} req
+ * @param {object} res
+ * @param {function} next
+ * @returns {Promise.<void>}
+ */
+controller.postSchoolImpersonation = async (req, res, next) => {
+  const { dfeNumber } = req.body
+  try {
+    const outcome = await schoolImpersonationService.validateCreateImpersonation(req.user, dfeNumber)
```
[the remark this answers is not in the document]
Comment (GeorgeChatzigiannis) [reply admin/controllers/helpdesk.js]: @jon-shipley done

File: admin/services/school-impersonation.service.js (hunk 5 above)
Comment (GeorgeChatzigiannis) [reply admin/services/school-impersonation.service.js]: @jon-shipley updated as per our earlier discussion

File: admin/services/school-impersonation.service.js (hunk 4 above)
Comment (GeorgeChatzigiannis) [reply admin/services/school-impersonation.service.js]: @jon-shipley the first validator checks if the value is non empty where the second checks if the dfeNumber from the client does match a school. 
In between these two validators there is a database transaction to match the school record.
The alternative would be to have one validator with two different methods.

File: admin/services/school-impersonation.service.js (hunk 3 above)
Comment (GeorgeChatzigiannis) [reply admin/services/school-impersonation.service.js]: @jon-shipley merged to a single validator

File: admin/services/school-impersonation.service.js (hunk 4 above)
Comment (GeorgeChatzigiannis) [reply admin/services/school-impersonation.service.js]: @jon-shipley merged to one as it seems to be cleaner.

File: admin/controllers/helpdesk.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (jon-shipley) [reply admin/controllers/helpdesk.js]: Perhaps I wasn't clear.  You are changing state on the results of a GET which shouldn't happen.  You should do a POST to change state.

File: admin/controllers/helpdesk.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    const outcome = await schoolImpersonationService.validateImpersonationForm(req.user, dfeNumber)
+    if (outcome.hasError && outcome.hasError()) {
+      return controller.getSchoolImpersonation(req, res, next, outcome)
+    }
+    res.redirect('home')
+  } catch (error) {
+    next(error)
+  }
+  return next()
+}
```
Comment (jon-shipley) [inline admin/controllers/helpdesk.js]: Reading this I can't see where any work actually gets done.

File: admin/services/data-access/user.data.service.js [hunk 12]
```diff
@@ -24,6 +24,36 @@ const userDataService = {
...
+    const sql = `
+      SELECT
+        u.id,
+        u.passwordHash,
+        u.identifier,
+        r.title AS roleName,
+        u.school_id AS schoolId,
+        s.dfeNumber,
+        sce.timezone
+      FROM mtc_admin.${table} u
```
Comment (jon-shipley) [inline admin/services/data-access/user.data.service.js]: We have decided to drop this variable and use the string of the table name instead.

File: admin/services/school-impersonation.service.js [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+schoolImpersonationService.validateImpersonationForm = async (user, dfeNumber) => {
+  let validationError
+  validationError = schoolImpersonationDfeNumberValidator.isDfeNumberEmpty(dfeNumber)
+  if (validationError.hasError()) {
+    return validationError
+  }
+  let school
+  try {
+    school = await schoolDataService.sqlFindOneByDfeNumber(dfeNumber)
+  } catch (error) {}
```
Comment (jon-shipley) [inline admin/services/school-impersonation.service.js]: What do you think to naming it `ignore` where the catch error variable is unused?

File: admin/lib/validator/school-impersonation-dfe-number-validator.js [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+  if (!dfeNumber) {
+    validationError.addError('dfeNumber', SchoolImpersonationMessages.noInput)
+    return validationError
+  }
+  return validationError
+}
+
+const isDfeNumberValid = (school) => {
+  let validationError = new ValidationError()
+  if (!school || !school.id) {
```
Comment (jon-shipley) [inline admin/lib/validator/school-impersonation-dfe-number-validator.js]: `{
  id: 'foo'
}`

passes validation but is not valid.

File: admin/services/school-impersonation.service.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+  validationError = schoolImpersonationDfeNumberValidator.isDfeNumberValid(school)
```
Comment (jon-shipley) [inline admin/services/school-impersonation.service.js]: Not sure what the point of this statement is?  Either school is undefined from line 21, or it was found in the db.  Why do we need a validator to detect undefined?  Just use `if (school === undefined)`

File: admin/services/school-impersonation.service.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+  if (validationError.hasError()) {
+    return validationError
+  }
+  user.School = school.dfeNumber
+  user.schoolId = school.id
+  user.timezone = school.timezone
```
Comment (jon-shipley) [inline admin/services/school-impersonation.service.js]: It seems to me that we lose the knowledge that the user is impersonating.  We should be able to know:
* that the user is impersonating
* the true role for the user

Can you confirm?

File: admin/services/school-impersonation.service.js [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+  }
+  let school
+  try {
+    school = await schoolDataService.sqlFindOneByDfeNumber(dfeNumber)
+  } catch (error) {}
+  validationError = schoolImpersonationDfeNumberValidator.isDfeNumberValid(school)
+  if (validationError.hasError()) {
+    return validationError
+  }
+  user.School = school.dfeNumber
```
Comment (jon-shipley) [inline admin/services/school-impersonation.service.js]: I don't think that a  function named `validateImpersonationForm` should affect the impersonation.  That would be more likely named `impersonate` or something similar.

File: admin/services/school-impersonation.service.js [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+  return user
+}
+
+/**
+ * Remove impersonation data from request.user object
+ * @param {object} user
+ * @returns {void}
+ */
+schoolImpersonationService.removeImpersonation = (user) => {
+  R.forEach(k => delete user[k], ['School', 'schoolId', 'timezone'])
```
Comment (jon-shipley) [inline admin/services/school-impersonation.service.js]: Should we add a message to the user?

File: admin/views/school/school-home.ejs [hunk 19]
```diff
@@ -1,6 +1,16 @@
+
 <div class="govuk-grid-row">
     <div class="govuk-grid-column-full">
-        <%- partial('partials/_name_and_sign_out') %>
+        <% if (user.role === 'HELPDESK') { %>
```
Comment (jon-shipley) [inline admin/views/school/school-home.ejs]: Might be better to remove this magic string.  Something as important as role names might be kept in a frozen object literal.

File: admin/views/school/school-home.ejs [hunk 20]
```diff
@@ -1,6 +1,16 @@
...
+            <div id="sign-out">
+                <a class="govuk-link" href="/helpdesk/school-impersonation">Remove Impersonation</a>
+            </div>
+            <% if (typeof user != 'undefined') { %>
```
Comment (jon-shipley) [inline admin/views/school/school-home.ejs]:
```suggestion
            <% if (typeof user !== 'undefined') { %>
```

**Comment by GuyHarwood** [comment]
@activemq @sinhaamit1980 this is not ready for testing yet as its not dev complete

File: admin/controllers/helpdesk.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/controllers/helpdesk.js]: also - what if they want to remain as that school?  the removal of impersonation needs to be explicit.

File: admin/controllers/helpdesk.js [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+
+/**
+ * Renders the helpdesk (role) landing page
+ * @param {object} req
+ * @param {object} res
+ * @param {function} next
+ * @returns {Promise.<void>}
+ */
+controller.getSchoolLandingPage = async (req, res, next) => {
```
Comment (GuyHarwood) [inline admin/controllers/helpdesk.js]: i thought the duplicate of the school home page was being removed in favour of reusing the existing school home page? or is this a different page?

File: admin/controllers/helpdesk.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (GeorgeChatzigiannis) [reply admin/controllers/helpdesk.js]: @jon-shipley @GuyHarwood considering this indeed should have been an isolated action this has now changed and is an explicit POST route.

File: admin/spec/back-end/lib/validator/school-impersonation/school-impersonation-empty-value-validator.spec.js (hunk 6 above)
Comment (GeorgeChatzigiannis) [reply admin/spec/back-end/lib/validator/school-impersonation/school-impersonation-empty-value-validator.spec.js]: @jon-shipley resolving this as the caller service implementation and naming has changed.

File: admin/controllers/helpdesk.js (hunk 11 above)
Comment (GeorgeChatzigiannis) [reply admin/controllers/helpdesk.js]: @jon-shipley changed as described on the other comment

File: admin/services/data-access/user.data.service.js (hunk 12 above)
Comment (GeorgeChatzigiannis) [reply admin/services/data-access/user.data.service.js]: @jon-shipley updated to use the table name.

File: admin/services/school-impersonation.service.js (hunk 13 above)
Comment (GeorgeChatzigiannis) [reply admin/services/school-impersonation.service.js]: @jon-shipley as discussed this is updated.

File: admin/lib/validator/school-impersonation-dfe-number-validator.js (hunk 14 above)
Comment (GeorgeChatzigiannis) [reply admin/lib/validator/school-impersonation-dfe-number-validator.js]: @jon-shipley at that stage the school either has been fetched from the database or it doesn't exist at all.if a 'foo' value is stored as id that would be fundamentally wrong.

File: admin/services/school-impersonation.service.js (hunk 16 above)
Comment (GeorgeChatzigiannis) [reply admin/services/school-impersonation.service.js]: @jon-shipley knowledge so far about if the user is already impersonating is not needed and the role itself is exposed already in the authentication part.

File: admin/services/school-impersonation.service.js (hunk 17 above)
Comment (GeorgeChatzigiannis) [reply admin/services/school-impersonation.service.js]: @jon-shipley changed as per discussion.

File: admin/views/school/school-home.ejs (hunk 19 above)
Comment (GeorgeChatzigiannis) [reply admin/views/school/school-home.ejs]: @jon-shipley replaced as per our discussion and your feedback.

File: admin/controllers/helpdesk.js (hunk 21 above)
Comment (GeorgeChatzigiannis) [reply admin/controllers/helpdesk.js]: @GuyHarwood this has been updated to use school controller.

File: admin/authentication/middleware.js [hunk 22]
```diff
@@ -2,18 +2,22 @@
...
+      logger.debug(`checking authorisation on ${req.url} for role:${roles} against userRole:${userRole}`)
+      // declare single role variable if input is not array of roles
+      const role = !Array.isArray(roles) && roles
+      if (role === undefined ||
+        (role !== undefined && userRole && role === userRole) ||
+        (Array.isArray(roles) && roles.some(r => r === userRole))) {
         return next()
       } else {
-        logger.warn(`could not authorise ${role} against userRole:${userRole}, UserName:${req.user.UserName} ID:${req.user.id}`)
+        logger.warn(`could not authorise ${roles} against userRole:${userRole}, UserName:${req.user.UserName} ID:${req.user.id}`)
```
Comment (GuyHarwood) [inline admin/authentication/middleware.js]: how does `roles` render if it's an array?

File: admin/services/helpdesk.service.js [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+
+const roles = require('../lib/consts/service-roles.js')
+const helpdeskService = {}
+
+/**
+ * Identify if helpdesk user has not received impersonation
+ * @param {object} user
+ * @returns {boolean}
+ */
+helpdeskService.hasHelpdeskNotReceivedImpersonation = (user) => {
```
Comment (GuyHarwood) [inline admin/services/helpdesk.service.js]: that method name doesn't read very well.  maybe flip it to `isImpersonatingSchool()` ?

File: admin/services/school-impersonation.service.js [hunk 24]
```diff
@@ -12,24 +12,29 @@ const schoolImpersonationService = {}
  * @param {string} dfeNumber
  * @returns {Object}
  */
-schoolImpersonationService.validateImpersonationForm = async (user, dfeNumber) => {
+schoolImpersonationService.processImpersonationForm = async (user, dfeNumber) => {
```
Comment (GuyHarwood) [inline admin/services/school-impersonation.service.js]: again, odd choice of method name.  maybe `setSchoolImpersonation()` ?

File: admin/controllers/school.js [hunk 25]
```diff
@@ -20,6 +21,11 @@ const controller = {}
 controller.getSchoolLandingPage = async (req, res, next) => {
   res.locals.pageTitle = 'School Homepage'
   try {
+    const hasHelpdeskNotReceivedImpersonation = helpdeskService.hasHelpdeskNotReceivedImpersonation(req.user)
```
Comment (jon-shipley) [inline admin/controllers/school.js]: I think naming things in the negative is generally harder to understand.  Would be better as `!isHelpdeskImpersonator()` IMHO.  Happy to discuss.

File: admin/lib/consts/service-roles.js [hunk 26]
```diff
@@ -0,0 +1,6 @@
+module.exports = Object.freeze({
+  helpdesk: 'HELPDESK',
+  serviceManager: 'TEACHER',
+  teacher: 'TEACHER',
+  testDeveloper: 'TEST-DEVELOPER'
+})
```
Comment (jon-shipley) [inline admin/lib/consts/service-roles.js]: Whilst it is much better having the object 👍 I think the serviceManager role needs to be set to SERVICE-MANAGER 😉

File: admin/roles-config.js [hunk 27]
```diff
@@ -1,8 +1,10 @@
 module.exports = {
   ROLE_SERVICE_MANAGER: 'SERVICE-MANAGER',
   ROLE_TEST_DEVELOPER: 'TEST-DEVELOPER',
+  ROLE_HELPDESK: 'HELPDESK',
```
Comment (jon-shipley) [inline admin/roles-config.js]: Not sure what this object is doing.  It's a mess - some sort of combination of roles and home pages? 😱

File: admin/routes/helpdesk.js [hunk 28]
```diff
@@ -0,0 +1,15 @@
+'use strict'
+
+const express = require('express')
+const router = express.Router()
+const isAuthenticated = require('../authentication/middleware')
+const rolesConfig = require('../roles-config')
+const helpdeskController = require('../controllers/helpdesk')
+
+router.get('/', isAuthenticated(rolesConfig.ROLE_HELPDESK), (req, res, next) => helpdeskController.getSchoolImpersonation(req, res, next))
```
Comment (jon-shipley) [inline admin/routes/helpdesk.js]: Shouldn't we be using the new roles.js object that defines all the roles in the system?  Prefix with ROLE seems superfluous to me.

File: admin/lib/consts/service-roles.js [hunk 29]
```diff
@@ -0,0 +1,6 @@
+module.exports = Object.freeze({
```
Comment (jon-shipley) [inline admin/lib/consts/service-roles.js]: This files should just be named roles.js - it should be all roles in the system.

File: admin/services/school-impersonation.service.js (hunk 18 above)
Comment (jon-shipley) [reply admin/services/school-impersonation.service.js]: @GeorgeChatzigiannis - do you think it is worth adding a flash message for the user?

File: admin/views/helpdesk/school-impersonation.ejs [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+                                    <li><a href="#<%= f %>"><%= errors %></a></li>
+                                <% } %>
+                            <% }) %>
+                        </ul>
+                    </div>
+                </div>
+            <% } %>
+            <h1 class="govuk-heading-xl"><%= pageTitle %></h1>
+            <%- partial('partials/_gds_session_expiry') %>
+            <form action="/helpdesk/add-school-impersonation" id="schoolImpersonationForm" method="post">
```
Comment (jon-shipley) [inline admin/views/helpdesk/school-impersonation.ejs]: I think it would be better to prevent browsers caching dfeNumbers here.  In normal operation, this could be used hundreds of times a day.  I think it would be safer to make the user type out the whole number.

`autocomplete=off`

File: admin/roles-config.js (hunk 27 above)
Comment (GuyHarwood) [reply admin/roles-config.js]: this is one thing i forgot to bring up - there should be just one canonical set of declared roles in the admin app

File: admin/authentication/middleware.js (hunk 22 above)
Comment (GeorgeChatzigiannis) [reply admin/authentication/middleware.js]: @GuyHarwood changed so that it displays the message correctly regardless of string or array type.

File: admin/authentication/middleware.js [hunk 31]
```diff
@@ -2,18 +2,22 @@
...
     if (req.isAuthenticated()) {
       let userRole
       if (req.user) {
         userRole = ((req.user).role || {})
       }
-      logger.debug(`checking authorisation on ${req.url} for role:${role} against userRole:${userRole}`)
-      if (role === undefined || (role !== undefined && userRole && role === userRole)) {
+      logger.debug(`checking authorisation on ${req.url} for role:${roles} against userRole:${userRole}`)
+      // declare single role variable if input is not array of roles
+      const role = !Array.isArray(roles) && roles
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/authentication/middleware.js]: also - why not add tests around this?  its a critical part of the system

File: admin/services/school-impersonation.service.js (hunk 24 above)
Comment (GeorgeChatzigiannis) [reply admin/services/school-impersonation.service.js]: @GuyHarwood updated

File: admin/services/helpdesk.service.js (hunk 23 above)
Comment (GeorgeChatzigiannis) [reply admin/services/helpdesk.service.js]: @GuyHarwood this method has now been split to 2 different ones for readability and simplification.

File: admin/services/helpdesk.service.js [hunk 32]
```diff
@@ -4,17 +4,21 @@ const roles = require('../lib/consts/service-roles.js')
...
+  return user.role === roles.helpdesk
+}
+
+/**
+ * Identify if current user has impersonation properties set
+ * @param {object} user
+ * @returns {boolean}
+ */
+helpdeskService.isImpersonating = (user) => {
+  return user.School && user.schoolId && typeof user.timezone === 'string'
```
Comment (GuyHarwood) [inline admin/services/helpdesk.service.js]: i would just do a typeof check for number on schoolId.  its more explicit

File: admin/controllers/school.js (hunk 25 above)
Comment (GeorgeChatzigiannis) [reply admin/controllers/school.js]: @jon-shipley this is now simplified and split to two methods.

File: admin/lib/consts/service-roles.js (hunk 26 above)
Comment (GeorgeChatzigiannis) [reply admin/lib/consts/service-roles.js]: oh wow that was a good one. resolved!

File: admin/routes/helpdesk.js (hunk 28 above)
Comment (GeorgeChatzigiannis) [reply admin/routes/helpdesk.js]: @jon-shipley I could tackle that on the follow up PBI. seems like this piece of work is already doing too much.

File: admin/lib/consts/service-roles.js (hunk 29 above)
Comment (GeorgeChatzigiannis) [reply admin/lib/consts/service-roles.js]: @jon-shipley done

File: admin/views/helpdesk/school-impersonation.ejs (hunk 30 above)
Comment (GeorgeChatzigiannis) [reply admin/views/helpdesk/school-impersonation.ejs]: @jon-shipley that's a fair point. I have added the attribute.

File: admin/services/helpdesk.service.js (hunk 32 above)
Comment (GeorgeChatzigiannis) [reply admin/services/helpdesk.service.js]: @GuyHarwood fair point. amended

File: admin/services/school-impersonation.service.js (hunk 18 above)
Comment (GeorgeChatzigiannis) [reply admin/services/school-impersonation.service.js]: @jon-shipley it's already included. Please have a look at the helpdesk controller method `postRemoveSchoolImpersonation`.

File: admin/roles-config.js (hunk 27 above)
Comment (GeorgeChatzigiannis) [reply admin/roles-config.js]: @GuyHarwood @jon-shipley already made a note for the follow-up about this.

File: admin/authentication/middleware.js (hunk 31 above)
[the remark this answers is not in the document]
Comment (GeorgeChatzigiannis) [reply admin/authentication/middleware.js]: @GuyHarwood @jon-shipley please have another look on this.

File: admin/authentication/middleware.js [hunk 33]
```diff
@@ -2,18 +2,20 @@
...
+      let roles = typeof arg === 'string' && arg.length > 1
+        ? [arg] : arg
       let userRole
       if (req.user) {
         userRole = ((req.user).role || {})
       }
-      logger.debug(`checking authorisation on ${req.url} for role:${role} against userRole:${userRole}`)
-      if (role === undefined || (role !== undefined && userRole && role === userRole)) {
+      logger.debug(`checking authorisation on ${req.url} for roles:${roles && roles.join()} against userRole:${userRole}`)
+      if (roles === undefined || (Array.isArray(roles) && roles.some(r => r === userRole))) {
```
Comment (jon-shipley) [inline admin/authentication/middleware.js]: roles will now always be an array so this can be simplified.  userRole should be a valid role or undefined.

File: admin/controllers/school.js [hunk 34]
```diff
@@ -20,6 +21,12 @@ const controller = {}
 controller.getSchoolLandingPage = async (req, res, next) => {
   res.locals.pageTitle = 'School Homepage'
   try {
+    const isHelpdeskRole = helpdeskService.isHelpdeskRole(req.user)
+    const isImpersonating = helpdeskService.isImpersonating(req.user)
+    // Prevent helpdesk users from accessing school home when impersonation data are not populated
+    if (isHelpdeskRole && !isImpersonating) {
```
Comment (jon-shipley) [inline admin/controllers/school.js]: 👍 clear

File: admin/lib/consts/roles.js [hunk 35]
```diff
@@ -0,0 +1,6 @@
+module.exports = Object.freeze({
+  helpdesk: 'HELPDESK',
+  serviceManager: 'SERVICE-MANAGER',
+  teacher: 'TEACHER',
+  testDeveloper: 'TEST-DEVELOPER'
+})
```
Comment (jon-shipley) [inline admin/lib/consts/roles.js]: 👍

File: admin/services/school-impersonation.service.js [hunk 36]
```diff
@@ -0,0 +1,49 @@
...
+const schoolImpersonationService = {}
+
+/**
+ * Validate and assign school to helpdesk session
+ * @param {object} user
+ * @param {string} dfeNumber
+ * @returns {Object}
+ */
+schoolImpersonationService.setSchoolImpersonation = async (user, dfeNumber) => {
+  let validationError
```
Comment (jon-shipley) [inline admin/services/school-impersonation.service.js]: Why can't this be const?

File: admin/services/school-impersonation.service.js [hunk 37]
```diff
@@ -0,0 +1,49 @@
...
+  // returns a validation error if dfeNumber provided is empty
+  if (validationError.hasError()) {
+    return validationError
+  }
+  let school
+  try {
+    school = await schoolDataService.sqlFindOneByDfeNumber(dfeNumber)
+  } catch (ignore) {}
+  // returns a validation error if the school record is not valid
+  validationError = schoolImpersonationValidator.isSchoolRecordValid(school)
```
Comment (jon-shipley) [inline admin/services/school-impersonation.service.js]: This is another variable, e.g. schoolValidationError.  Should be const.

File: admin/spec/back-end/authentication/middleware.spec.js [hunk 38]
```diff
@@ -0,0 +1,110 @@
...
+      const req = httpMocks.createRequest(reqParams)
+      spyOn(req, 'isAuthenticated').and.returnValue(true)
+      spyOn(res, 'redirect')
+      const allowedRoles = ['SERVICE-MANAGER', 'TEACHER']
+      const func = isAuthenticated(allowedRoles)
+      func(req, res, next)
+      expect(next).not.toHaveBeenCalled()
+      expect(res.redirect).toHaveBeenCalledWith('/unauthorised')
+    })
+    it('calls next if the role passed is undefined while the request is authenticated', () => {
```
Comment (jon-shipley) [inline admin/spec/back-end/authentication/middleware.spec.js]: Also need same test for null, empty string, function, and object types.

File: admin/spec/back-end/authentication/middleware.spec.js [hunk 39]
```diff
@@ -0,0 +1,110 @@
...
+      const req = httpMocks.createRequest(reqParams)
+      spyOn(req, 'isAuthenticated').and.returnValue(false)
+      spyOn(res, 'redirect')
+      const func = isAuthenticated(undefined)
+      func(req, res, next)
+      expect(next).not.toHaveBeenCalled()
+      expect(res.redirect).toHaveBeenCalledWith('/sign-in')
+    })
+  })
+})
```
Comment (jon-shipley) [inline admin/spec/back-end/authentication/middleware.spec.js]: Should be a test where the userRole is not defined.

File: admin/authentication/middleware.js [hunk 40]
```diff
@@ -2,18 +2,20 @@
 
 const logger = require('../services/log.service').getLogger()
 
-function isAuthenticated (role) {
+function isAuthenticated (arg) {
   return function (req, res, next) {
     if (req.isAuthenticated()) {
+      let roles = typeof arg === 'string' && arg.length > 1
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/authentication/middleware.js]: Ramda?

File: admin/authentication/middleware.js (hunk 33 above)
Comment (GuyHarwood) [reply admin/authentication/middleware.js]: ramda contains?

File: admin/spec/back-end/authentication/local-strategy.spec.js [hunk 41]
```diff
@@ -0,0 +1,87 @@
...
+      spyOn(userDataService, 'sqlFindUserInfoByIdentifier')
+      spyOn(adminLogonEventDataService, 'sqlCreate')
+      const doneFunc = (err, res) => err || res
+      await validateAndSave(req, 'teacher1', 'password', doneFunc)
+      expect(adminLogonEventDataService.sqlCreate).toHaveBeenCalledWith(
+        { sessionId: 1, body: '{}', remoteIp: 'remoteAddress', userAgent: undefined, loginMethod: 'local', errorMsg: 'Invalid user', isAuthenticated: false }
+      )
+    })
+  })
+})
```
Comment (GuyHarwood) [inline admin/spec/back-end/authentication/local-strategy.spec.js]: 🏆

File: admin/services/school-impersonation.service.js (hunk 36 above)
Comment (GeorgeChatzigiannis) [reply admin/services/school-impersonation.service.js]: @jon-shipley updated

File: admin/services/school-impersonation.service.js (hunk 37 above)
Comment (GeorgeChatzigiannis) [reply admin/services/school-impersonation.service.js]: @jon-shipley updated

File: admin/spec/back-end/authentication/middleware.spec.js (hunk 39 above)
Comment (GeorgeChatzigiannis) [reply admin/spec/back-end/authentication/middleware.spec.js]: @jon-shipley done

File: admin/spec/back-end/authentication/middleware.spec.js (hunk 38 above)
Comment (GeorgeChatzigiannis) [reply admin/spec/back-end/authentication/middleware.spec.js]: @jon-shipley all done and are now redirecting instead of calling next.

File: admin/authentication/middleware.js [hunk 42]
```diff
@@ -2,18 +2,20 @@
...
+        ? [arg] : arg
       let userRole
       if (req.user) {
         userRole = ((req.user).role || {})
```
[the remark this answers is not in the document]
Comment (GeorgeChatzigiannis) [reply admin/authentication/middleware.js]: @jon-shipley that is correct. updated.

File: admin/authentication/middleware.js (hunk 40 above)
[the remark this answers is not in the document]
Comment (GeorgeChatzigiannis) [reply admin/authentication/middleware.js]: @jon-shipley updated so they are now triggering a redirect.

[the remark this answers is not in the document]
Comment (GeorgeChatzigiannis) [reply admin/authentication/middleware.js]: I prefer if it only checks for an array structure and construct an array if not since the following code checks are made on that assumption.

File: admin/authentication/middleware.js (hunk 33 above)
Comment (GeorgeChatzigiannis) [reply admin/authentication/middleware.js]: @jon-shipley the undefined check is now removed.

File: admin/spec/back-end/authentication/local-strategy.spec.js (hunk 41 above)
Comment (GeorgeChatzigiannis) [reply admin/spec/back-end/authentication/local-strategy.spec.js]: @GuyHarwood you mean ramda.includes ;)
